### ISSUE TITLE
XmlSerializer Creation throws FileNotFoundException

### ISSUE THREAD

**Issue by david-maw** [dotnet/runtime#83152]
### Description

The unexpected pause does not do any harm in that the missing file seems to be optional and the exception gets caught before it returns to user code, Nonetheless it's mighty confusing and tedious to get around, especially on Android, because even if you disable break on FileNotFoundException the debugger still pauses with a message like this:

![nobreakdialog](https://user-images.githubusercontent.com/12423046/175719472-6873b046-c12f-43b0-a329-1987415d87cb.jpg)

Even more annoyingly it seems to break (or pause) twice for each new XmlSerializer created.

There are several things that don't seem right:

1. Using an exception to report a normal situation (a missing optional file) is likely to cause confusion and is especially annoying migrating code from Xamarin which does not do this.
2. Throwing the exception once per XmlSerializer is bad enough, doing it twice is worse
3. In Android disabling that Break applies to all modules (unlike Windows where you can at least ignore exceptions in a named module). This may be a feature of Android.
4. When you DO inhibit the break in Android it pauses anyway (with the message above). 

If this needs to be split into multiple issues, I can do that, just let me know.  

### Steps to Reproduce

1. Unzip the project from [XmlSerializerException.zip](https://github.com/dotnet/maui/files/8982430/XmlSerializerException.zip)
2. In visual Studio Exception Settings enable Break on `System.IO.FileNotFoundException`
3.  Build and run the project on Android, it should break on the statement 
    `new XmlSerializer(typeof(Person)); ` 
    in MainPage.Xaml.cs
4. Press "continue" and it will break again.
5. Restart the application
6. This time disable Break on `System.IO.FileNotFoundException`
7. Run the project under the debugger again and note that it still pauses on  `System.IO.FileNotFoundException` even though Break is disabled for that Exception.  

### Version with bug

6.0.400 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

Android 11, Windows 10 

### Did you find any workaround?

Just continue through it and the system will usually keep working.

**Comment by david-maw** [comment]
No idea what changed but with .NET 7 Android seems ok and Windows shows the problem.

**Comment by mfeingol** [comment]
I'm still seeing this with .NET 7 Maui and Android.

**Comment by StephenMolloy** [comment]
I suspect when this issue does not present itself, it is because XmlSerializer is using a reflection-based implementation under the covers which does not attempt to load a pre-generated serializer assembly.

**Comment by SaqiiiCreates** [comment]
Still experiencing this issue on .NET 7 with android and windows.

**Comment by rgroenewoudt** [comment]
What about making XmlSerializer.Mode public to force reflection mode? This would prevent loading the pre-gen assemblies.

**Comment by jeremy-visionaid** [comment]
Curiously I still get the FileNotFoundException even when the XmlSerializers.dll files exists when using the serializer generator:
https://learn.microsoft.com/en-us/dotnet/core/additional-tools/xml-serializer-generator

**Comment by exedor** [comment]
I'm getting this exception as well upon first request with a WebAPI project. I am using SignalRSwaggerGen for documenting a SignalR Hub with Swagger API doc gen. Here is the exception thrown. There are two exceptions related to this, both mostly the same thrown at two different points in the call stack:

System.IO.FileNotFoundException
  HResult=0x80070002
  Message=Could not load file or assembly 'SignalRSwaggerGen.XmlSerializers, Version=4.5.0.0, Culture=neutral, PublicKeyToken=57a441116a7c7fc1, processorArchitecture=MSIL'. The system cannot find the file specified.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>

I'm currently trying to figure out how to deal with this. I'd prefer not to use the "Only my code" solution because I'm learning sometimes it is the way my code is interacting with core libs or 3rd party assemblies that causes exceptions and find that to sometimes be valuable indicators that something could be getting misused.

**Comment by limeniye** [comment]
I have this problem in my IIncrementalGenerator.
It occurs when I try to rebuild the project second time.

**Comment by stewienj** [comment]
> Curiously I still get the FileNotFoundException even when the XmlSerializers.dll files exists when using the serializer generator: https://learn.microsoft.com/en-us/dotnet/core/additional-tools/xml-serializer-generator

@jeremy-visionaid I had the same issue, had to do my own AssemblyResolve handler, as seen in my working example here:
https://github.com/stewienj/StewienCodeExamples/tree/master/SerializerSucess

I would welcome an alternative solution.

**Comment by jeremy-visionaid** [comment]
@stewienj Awesome, thanks for sharing! Fingers crossed for an upstream fix still :)

**Comment by rokmeglic71** [comment]
This happens to us to. Using .NET 9.0, latest MAUI 9.0.30.
Error: FileNotFoundException: IO_FileName_Name, XmlSerializers

[CoPilot AI](https://copilot.microsoft.com/chats/PRKPnFATuqX6ZE1smqLPs) redirect me to:
 - [C# XmlSerializer FileNotFound exception, By rickvdbosch](https://www.rickvandenbosch.net/blog/c-xmlserializer-filenotfound-exception/)
 - [XmlSerializer throws FileNotFoundException, Posted by [Manuel]](https://manuelmeyer.net/2012/07/xmlserializer-throws-filenotfoundexception/)

**Extra info:**
- [StackOverFlow answers](https://stackoverflow.com/questions/1127431/xmlserializer-giving-filenotfoundexception-at-constructor)

**Comment by limeniye** [comment]
I have some old code in my project that resolves these errors for my generator. It might be useful to someone:

```C#
internal sealed class AssemblyResolver : IDisposable
{
    private readonly ConcurrentDictionary<string, Assembly> _forcedLoadedAssemblies = new ConcurrentDictionary<string, Assembly>();
    private readonly ResolveEventHandler _resolveEventHandler;
    private bool _isRegistered = false;

    public AssemblyResolver(params string[] assembliesToLoad)
    {
        _resolveEventHandler = new ResolveEventHandler(OnCurrentDomainAssemblyResolve);
        Register(assembliesToLoad);
    }

    private void Register(string[] assembliesToLoad)
    {
        if (!_isRegistered)
        {
            AppDomain.CurrentDomain.AssemblyResolve += _resolveEventHandler;

            foreach (var assemblyName in assembliesToLoad)
            {
                LoadAssembly(assemblyName);
            }

            _isRegistered = true;
        }
    }

    private void LoadAssembly(string assemblyName)
    {
        string assemblyPath = RuntimeEnvironment.GetRuntimeDirectory();
        string assemblyFullPath = Path.Combine(assemblyPath, $"{assemblyName}.dll");

        var loadedAssembly = Assembly.LoadFrom(assemblyFullPath);
        _forcedLoadedAssemblies.GetOrAdd(assemblyName, loadedAssembly);
    }

    private Assembly OnCurrentDomainAssemblyResolve(object sender, ResolveEventArgs args)
    {
        var name = new AssemblyName(args.Name).Name;
        _forcedLoadedAssemblies.TryGetValue(name, out Assembly outAssembly);
        return outAssembly;
    }

    public void Dispose()
    {
        if (_isRegistered)
        {
            AppDomain.CurrentDomain.AssemblyResolve -= _resolveEventHandler;
            _isRegistered = false;
        }
    }
}

```

A simple example:
```C#
[Generator]
public class MyIncrementalGenerator : IIncrementalGenerator, IDisposable
{
    private readonly AssemblyResolver _assemblyResolver;

    public MyIncrementalGenerator()
    {
        _assemblyResolver = new AssemblyResolver("System.ComponentModel.Annotations");
    }
}
```

**Comment by StephenMolloy** [comment]
Flipping `XmlSerializer` between ILGen-based and Reflection-based modes is not a feature we want to expose publicly. The ILGen-based implementation is more complete and is thus the default... but it's messy and limited in many ways and we would like to find a path towards relying on the reflection-based implementation in the future. But toggling between the two and deciding which one to use is internal.

As for the exceptions being thrown - those are out of `XmlSerializer`'s control. We need to check to see if a pre-generated serializer already exists, and this (`Assembly.Load`) is one of those cases that is a notable exception to the "don't use exceptions for control flow" rule. Here's a quick rundown from Copilot who is better than me with words.

---

## **Exceptions as Control Flow in .NET 6+: A Practical Perspective**

The .NET design guidelines emphasize that exceptions should be reserved for truly exceptional conditions, not for routine control flow. This principle is outlined in the official documentation:  
> “Do not use exceptions for normal flow of control, as they can be expensive in terms of performance.”  
— https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/exception-throwing

However, in practice, there are scenarios in .NET 6 and later where exceptions are used as part of normal operational logic—either because no alternative exists or because the exception-based model is the most practical and robust solution. Two notable examples include:

---

### **1. Assembly Loading with `Assembly.Load()`**

When dynamically loading assemblies, such as in plugin systems or optional feature discovery, the only reliable way to determine if an assembly is available is to attempt loading it and catch any exceptions that occur. There is no `TryLoad` API in .NET, and pre-checking for file existence or probing manually is error-prone and race-condition-prone.

```csharp
try
{
    var asm = Assembly.Load("Optional.Plugin");
    // Proceed with using the assembly
}
catch (FileNotFoundException)
{
    // Plugin not available — fallback gracefully
}
```

This pattern is widely used and implicitly accepted in the .NET ecosystem. Even though it technically violates the “no exceptions for control flow” rule, it is considered a necessary and pragmatic exception. The .NET runtime itself uses similar exception-based probing internally during assembly resolution.

**Reference:**  
- [[Assembly.Load Method - Microsoft Docs](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-9.0)](https://learn.microsoft.com/en-us/dotnet/api/system.reflection.assembly.load?view=net-9.0)

---

### **2. Task Cancellation via `OperationCanceledException`**

In asynchronous programming, cancellation is a common and expected outcome. The standard pattern in .NET is to throw an `OperationCanceledException` when a `CancellationToken` is triggered. This exception is not treated as a failure but as a signal to unwind the call stack and exit the operation cleanly.

```csharp
cancellationToken.ThrowIfCancellationRequested();
```

This is a deliberate use of exceptions for control flow. The .NET runtime and libraries treat this exception differently from others—it marks a task as *Canceled*, not *Faulted*, and is often silently handled by the task infrastructure.

**Reference:**  
- https://learn.microsoft.com/en-us/dotnet/standard/threading/cancellation-in-managed-threads

---

### **Conclusion**

While the .NET guidelines discourage using exceptions for control flow, there are legitimate cases—especially in dynamic assembly loading and task cancellation—where exceptions are the only or best mechanism available. These patterns are used in the .NET runtime and supported by the framework’s design. Therefore, using `Assembly.Load()` with exception handling to detect optional assemblies is consistent with established practice and should not be considered a misuse of exceptions.

Would you like this formatted as a Markdown snippet or included in a draft response to the feature request?

**Comment by david-maw** [comment]
In the `Assembly. Load` case there's an appeal to pragmatism, but a pragmatic approach might well be to check the fairly simple case of "file not present" and if it is not, return an error. If it is present, go ahead and try to load it, accepting that if the load fails an exception will still be thrown (that's the branch that's race condition and error prone, I suspect).

That said, this answer comes down to "works as designed" or possibly "works as implemented" if I'm being cynical, so that's that. If it were easier to simply tell the debugger to ignore these exceptions life would have been simpler and that's a separate bug which is (I think) resolved now.

**Comment by jeremy-visionaid** [comment]
I've got to agree. This is another one of those death by 1000 cuts issues, where the current behavior here unnecessarily diminishes the developement experience. Consider that the VS Code debugging experience is clunky for ios/maccatalyst (https://github.com/microsoft/vscode-dotnettools/issues/1536). Consider that the WinUI app break point experience is unhelpful (https://github.com/microsoft/WindowsAppSDK/issues/5441). When there are other parts of the ecosystem where the behavior is bad, this compounds on top. So, closing this as "As Designed" or "Won't Fix" feels inappropriate - especially when the response didn't address why it just couldn't be side-stepped for the main case with `File.Exists` (accepting that it might sometimes still fail). The _regular_ execution of the serializer shouldn't behave this badly, I don't think the fallout from that should be simply ingored.

**Comment by rgroenewoudt** [comment]
Perhaps we should open a feature request for `Assembly.TryLoad`

**Comment by StephenMolloy** [comment]
There are two places `XmlSerializer` commonly runs into exceptions here, the first time is when calling `Assembly.Load(AssemblyName)`, and the second is when calling `Assembly.LoadFile(string)`. The latter case could conceivably be mitigated with a `File.Exists` check, but the former case will always be called first anyway. And there's just no reasonable way for `XmlSerializer` to know whether the current assembly-binding logic will find an assembly with the given name or not. It's a well worn pattern with `Assembly.Load`. If a "TryLoad" alternative existed, it would make sense to use that. But this is our only realistic option right now for `XmlSerializer`. :/

**Comment by david-maw** [comment]
Thanks for the explanation. Obviously I'm missing something. Why isn't a `File.Exists` check feasible in the`Assembly.Load(AssemblyName)` case? I'm perfectly willing to believe it isn't, but since I'm not familiar with the code I don't understand why. If the answer is "go read the code" that's fine, just point me to it. I couldn't agree more that a  "TryLoad" alternative would be useful in some cases (like this one), especially given the horrors of figuring out where to look for an executable.

**Comment by StephenMolloy** [comment]
An `AssemblyName` isn't a path. It's a class who's string representation is something like "myAssembly, Version=1.0.1234.0, Culture=en-US, PublicKeyToken=b77a5c561934e089c, ProcessorArchitecture=msil". That's not a path to check. And the CLR has special logic for finding assemblies to load, either in the local directory or on a special path (or from the GAC in the old days). And that logic can further be overridden by application developers to ensure the runtime can find their intended assemblies.

**Comment by david-maw** [comment]
Ah, got it, thanks. I was fooled by the name, I'd forgotten that an `AssemblyName` an was a multi-part object, mercifully, I've also been able to forget most of what I knew about the GAC but that was what  meant by "the horrors of figuring out where to look for an executable". That seems like the general case, but in this specific case (searching for a precompiled XML serializer) are we looking for a named file (which is what the exception implies) or are there other searches too, and only after they fail do we try for a named file and then throw an exception when it's not found?

**Comment by StephenMolloy** [comment]
When checking for a precompiled Xml serializer, `XmlSerializer` first looks for one by `AssemblyName`, since that's the more reliable/flexible way to locate an assembly. If that fails (which means we've already hit the first exception now), then we make one guess at the file name of the assembly and try to load that - which is where the second exception comes from.

**Comment by david-maw** [comment]
Got it, funny how everyone else's problem seems easy until you understand it... Thanks for taking the time to explain. Annoying that `Assembly.Load` works that way, but that's life.

**Comment by jeremy-visionaid** [comment]
Thanks for taking the time to explain. That is unfortunate, but understandable.

**Comment by ghost** [comment]
@rgroenewoudt where we would open the feature request for "Assembly.TryLoad"?

**Comment by rgroenewoudt** [comment]
> [@rgroenewoudt](https://github.com/rgroenewoudt) where we would open the feature request for "Assembly.TryLoad"?

Simply an new issue here in runtime repository as a API proposal. Add some motivation with a github search for Assembly.Load & FileNotFoundException to indicate it the usecase is common enough.

**Comment by ghost** [comment]
@rgroenewoudt are you willing to have a call (zoom, ...) and we can do it together?
